### PR TITLE
Append editorconfig namingstyle options to the top of the workspace o…

### PR DESCRIPTION
**Pull request by jmarolf**
Previously if we encountered an editorconfig file it would override naming styles the user had set in Visual Studio.  With this change the editorconfig styles are appended on top of the Visual Studio styles ensuring that editorconfig will always override Visual Studio if the styles conflict, but allowing naming styles specified in Visual Studio to still be used if there is no conflict

### CODE REVIEW

**Comment by dpoeschl** [comment]
Tagging @dotnet/roslyn-ide for review.

**Comment by rchande** [comment]
I don't understand what this change is trying to do at all. Is it trying to make sure that naming styles declared in .editorconfig files get processed before the ones in the options dialog?

**Comment by rchande** [comment]
Can you make the commit message more descriptive or add some comments?

File: src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs [hunk 1]
```diff
@@ -77,19 +77,25 @@ public EditorConfigStorageLocation(string keyName, Func<string, object> parseVal
         public EditorConfigStorageLocation()
         {
             // If the user didn't pass a keyName assume we need to parse the entire dictionary
-            _tryParseDictionary = (dictionary, type) =>
+            _tryParseDictionary = (dictionary, type, underlyingOption) =>
```
Comment (rchande) [inline src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs]: Why is this stored as a delegate?

**Comment by jasonmalinowski** [comment]
@jmarolf Is this PR still active? And what is it trying to do?

File: src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs (hunk 1 above)
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs]: The user can pass in their own function to be used.  This is  only used in the case where the user does not supply one

---
In reply to: [96749579](https://github.com/dotnet/roslyn/pull/16510#discussion_r96749579) [](ancestors = 96749579)

**Comment by jmarolf** [comment]
@jasonmalinowski yep, updated description

**Comment by jmarolf** [comment]
@dotnet/roslyn-ide please review

File: src/Workspaces/Core/Portable/NamingStyles/Serialization/NamingStylePreferencesExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles
+{
+    internal static class NamingStylePreferencesExtensions
+    {
+        public static NamingStylePreferences AppendNamingStylePreferencesToFront(this NamingStylePreferences original, NamingStylePreferences newPreferences)
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/NamingStyles/Serialization/NamingStylePreferencesExtensions.cs]: Nit: Call `PrependNamingStylePreferences` instead?

File: src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs [hunk 3]
```diff
@@ -77,19 +77,28 @@ public EditorConfigStorageLocation(string keyName, Func<string, object> parseVal
...
-                        !result.SymbolSpecifications.Any())
+                    var editorconfigNamingStylePreferences = EditorConfigNamingStyleParser.GetNamingStylesFromDictionary(dictionary);
+                    
+                    if (!editorconfigNamingStylePreferences.NamingRules.Any() &&
+                        !editorconfigNamingStylePreferences.NamingStyles.Any() &&
+                        !editorconfigNamingStylePreferences.SymbolSpecifications.Any())
                     {
-                        return (result: result, succeeded: false);
+                        // We were not able to parse any rules from editorconfig, tell the caller that the parse failed
+                        return (result: editorconfigNamingStylePreferences, succeeded: false);
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs]: Doesn't this mean that we get no naming styles at all if there is an editor config, but it doesn't specify naming styles?  Shouldn't we instead get the workspace ones?

Comment (jmarolf) [inline src/Workspaces/Core/Portable/Options/EditorConfigStorageLocation.cs]: returning false here will cause us to fall back to the workspace options as seen [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Options/OptionServiceFactory.cs,191)

**Comment by jmarolf** [comment]
retest windows_debug_unit32_prtest please

**Comment by jmarolf** [comment]
retest windows_release_vs-integration_prtest please

**Pull request merged**